contracts/core/EntryPoint.sol

```diff
@@ -225,13 +225,12 @@ contract EntryPoint is IEntryPoint, StakeManager {
 
```

Comment:
Question related to line 129 above (can't comment there because it hasn't changed in this PR)

Potential DoS vector on aggregators:
If the aggregated signature validation fails for any op, we catch it and revert with `SignatureValidationFailed(address(aggregator))`.  We don't give the aggregator an opportunity to report the `opIndex`.
The aggregator may access wallet storage during validation, so let's assume an aggregator that does. What if an attacker sends a valid aggregated op, then frontruns with a transaction that invalidates it by changing a storage slot in the wallet?
If someone does this to non-aggregated validation, the bundler identifies it by opIndex and drops the bad op.  But if it happens during `aggregator.validateSignatures(ops, opa.signature)`, the bundler doesn't know what to drop. It'll have to drop the aggregator itself, censoring all the valid aggregated ops.

Should we change the `validateSignatures` interface to return `(success,opIndex)` instead of just reverting?  So if it returns `success==false` we revert with `SignatureValidationFailed(address(aggregator),opIndex)`.  If the aggregator knows who the culprit, it returns it in `opIndex`.  If the signature verification fails outside the context of any specific op, it returns `opIndex=-1` to indicate that.  Or is there a better way to do this?

Reply:
I don't think we can add "opIndex" to help here: BLS aggregation merely sum the signature points - no per-op info.
The bundler calls simulate (which calls wallet.validateUserOperation and aggregator.valdiateUserOpSignature, which internally calls wallet.getPublicKey) just before creating the bundle.
So the way to make this attack, the wallet has to make getPublicKey return one valid off-chain, and then different value on-chain.
both calls to getPublicKey appear in the same sequence (just after validateUserOp) and called by the same sender (aggregator)
Do you think you can detect a difference? 

The other alternative is make a better mechanism to prevent modification of public-key:

One option is to raise the public-key into the protocol, and drop the UserOp if it gets modified (currently it is a private protocol between the aggregator and wallet).

Another optioin is to move the public-key storage into the aggregator itself, and add "time-lock" mechanism.

However, both of these solutions require deeper understanding of the aggregation by the core protocol (not sure that it is that a bad thing: a "wallet-specific-publickey" or equivalent is a concept that will exist in any aggregation system.

Reply:
The way is indeed to change the result of getPublicKey, or of getAggregator.  And there's no need to detect the difference in order to perform the attack.  The attacker puts the op in mempool while it's still valid, and then invalidates it using another transaction with higher gas, which changes the result of getPublicKey.

I guess it should normally be caught by the 2nd simulation while the op is still alone, so it won't happen in a batch.  But we do lose one line of defense which we otherwise have.  opIndex is meant to catch ops that somehow slipped through the 2nd simulation.

I'm not sure about adding the complexity of moving the key into the aggregator.  And in some cases it can be a complex "key" - something that isn't just a key but multiple keys with roles, although in this case the aggregator needs to know its structure anyway since it's the one parsing it.

In the flow you described above, where aggregator.valdiateUserOpSignature internally calls wallet.getPublicKey, it would be nice if we could somehow reflect the internal state if signature verification fails.  If I understand correctly, upon signature failure it can't detect which of the public keys caused it to fail.  But if it could somehow tell the bundler what was the public key it got from each wallet, the bundler could compare it to what it saw before, and find the culprit.  I don't see an easy way to export so much information though.  If it becomes an issue, clients might need to trace the execution and catch these changes if they ever occur.

Reply:
- the client can pull all public-keys of the ops it put into a bundle,
- in case the execution fails (meaning, one op managed to to modify it after the 2nd validation, but before putting it on chain), the client will pull all public-keys again, and compare the lists